inductiva/_benchmarks/methods.py

```diff
@@ -0,0 +1,370 @@
+    Args:
+        machines_to_run (List[str]): List of machines to run the simulation on.
+        already_ran (Dict[str, List[Task]]): Dictionary with the resources as
+            keys and the tasks ran on them as values.
+        intended_replicas (int): Number of replicas to run on each resource.
+    Returns:
+        Dict[str, int]: Dictionary with the resources as keys and the number of
+            tasks that need to be run on them as values.
+        Example: { "n1-standard-4": 2, "n1-standard-8": 1}
+    """
+    already_ran = defaultdict(list, already_ran)
+    return {m: intended_replicas - len(already_ran[m]) for m in machines_to_run}
+
+
+def _print_info(already_ran: Dict[str, List[Task]], tasks_to_run: Dict[str,
+                                                                       int]):
+    """
+    Prints the information about the tasks that have already been run and the
+    tasks that need to be run.
+    """
+
+    print()
+    print(__("tasks_ran"))
+    for key, value in already_ran.items():
+        print(f"--->{key}: {len(value)}")
+    print()
+    print(__("tasks_to_run"))
+    for key, value in tasks_to_run.items():
+        print(f"--->{key}: {value}")
+    print()
+
+
+def _run_replica(simulator_args: Dict[str, Union[Callable, Any]],\
+                 input_args: Dict[str, Union[Callable,Any]],\
+                 resource: BaseMachineGroup,\
+                 simulator: Simulator,\
+                 input_files: str,\
+                 replica: int):
+    """Runs a replica of the benchmark.
+
+    Args:
+        simulator_args (Dict[str, Union[Callable, Any]]): Arguments for the
+            simulator.
+        input_args (Dict[str, Union[Callable, Any]]): Arguments for the
+            input files.
+        resource (BaseMachineGroup): Resource to run the benchmark on.
+        simulator (Simulator): Simulator to run.
+        input_files (str): Path to the input files.
+        replica (int): Number of the replica.
+    """
+    # Each replica can run with different arguments
+    simulator_args_current = _render_dict(simulator_args, resource)
+    input_args_current = _render_dict(input_args, resource)
+    print(f"Simulation arguments {simulator_args_current}")
+    # input_files either come from the template manager or
+    # are passed directly.
+    simulator_args_current["input_dir"] = input_files
+    simulator_args_current["on"] = resource
+
+    if input_args_current:
+        target_dir = f"benchmark_inputs/{resource.machine_type}_{replica}"
+        TemplateManager.render_dir(source_dir=input_files,
+                                   target_dir=target_dir,
+                                   overwrite=False,
+                                   **input_args)
+        simulator_args_current["input_dir"] = target_dir
+
+    _ = simulator.run(**simulator_args_current)
+
+
+def _run(project: Project,
+         input_files: str,
+         replicas: int,
+         machines: List[Dict[str, Any]],
```

Comment:
@Pbarbosa92 Isn't this supposed to be a list of strings?

Reply:
Yes. You are correct. This was an older annotation that i forgot to update.